### PR TITLE
Rewrite to support suspend over IO and mtl.

### PR DESCRIPTION
**Pull request by 1Jajen1**
Things that need to be done before this can be merged:
- [ ] ~Function generation. The old version relied on `Eval` for stacksafety but since that does not work with `suspend` generators I need to come up with something else. That code is really ugly tho and thus will take a bit.~ I am not doing that this pr. It will come back, but it takes more work and thought. This currently fails because of stackoverflows and missing laziness and its kind of hard to fix.
- [ ] Stacksafe shrinking. Currently `Gen.map` calls into `Flow.map` which afaik is not stacksafe. This is rarely a concern for generators but needs to be addressed anyway. I am fine with skipping that tho.
- [x] Testing, this version isn't well tested just yet and that needs to change.
  - [x] Test the suspend interop and the implications multishot brings
  - [x] Test the environment parameter and if local works as expected
- [x] Gen.subterms up to arity 3
- [x] Gen debug functions like sample, printTree etc
- [x] Api docs

The new encoding is strictly less powerful but much easier to use.

### CODE REVIEW

**Comment by nomisRev** [comment]
> Function generation. The old version relied on Eval for stacksafety but since that does not work with suspend generators I need to come up with something else. That code is really ugly tho and thus will take a bit.

How hard would it be to write an `Eval` that supports `suspend` 🤔 Even if it's just for the purpose of having it as internal here?

**Comment by 1Jajen1** [comment]
> How hard would it be to write an Eval that supports suspend thinking Even if it's just for the purpose of having it as internal here?

I actually don't think that is hard at all, just like with the `AndThenS` class. It would be even easier if I do that without memoization.

But I also discovered other problems with suspend there that reach a little further and require substantial changes to the api, so for now I just use `runBlocking` inside the function gen...

Also I am running into so many problems even with this implementation that I am considering skipping this and redoing it later because most if not all other things are done and working. The jvm really isn't nice to this style of programming, no gadts, no stacksafety, no ghc-generics equivalent... I need to find a better encoding...

**Comment by 1Jajen1** [comment]
So quick status update: Feature wise this is good to go. I'll skip function generation as that requires a pr of its own + new ideas.
What this branch still needs is also what the current version lacks: More tests, tho mostly for internal components and the existing generator tests make use of those and thus kind of imply that those components work which is why I am somewhat confident in slowly adding tests over time rather than trying to cover all of it at once.

I'll also leave this pr in a ready to be merged state so I won't block arrow as a whole when we get to the next release^^

Documentation is the next task, for now I'll add a minimum viable bit of docs in the readme, but eventually this needs a few sections explaining property tests and some of the more hidden parts.

Lastly: Most of the deprecation messages currently come from the current parser combinators and the diffing algorithm which is based on recursion schemes. If those are a problem for the next release I can specialize both of those to unkinded versions.

**Comment by nomisRev** [comment]
Super awesome work @1Jajen1 👏 👏 👏 All your suggestions sounds perfect for me!
Will try to make some time tomorrow to review but it's not going to be easy 😅 `+3,258 −4,950`

**Comment by 1Jajen1** [comment]
> Will try to make some time tomorrow to review but it's not going to be easy sweat_smile `+3,258 −4,950`

Ah ye, it's quite large :see_no_evil: . I'd suggest taking a look at the encodings of `Gen`/`Rose` and `Test`/`PropertyTest` first and then view the rest of the it because without understanding what those do its going to be hard to parse some of the more in-depth methods.

File: arrow-check/src/main/kotlin/arrow/check/internal/AndThenS.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package arrow.check.internal
+
+// Add to arrow core?
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/internal/AndThenS.kt]: I think we should once we can inherit from `suspend (I) -> O`

File: arrow-check/src/test/kotlin/arrow/check/RunnerTest.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    // test interleaved suspension
+    "Interleaved delay on a failed test"(property {
+        // This validates interleaved suspension even on multishot uses during shrinking
+        val res = check {
+            val xs = forAll(Gen.int(0..100))
+            delay(10)
+            assert(xs <= 10)
+        }.not()
+
+        assert(res) { "Test did not fail".text() }
```
Comment (nomisRev) [inline arrow-check/src/test/kotlin/arrow/check/RunnerTest.kt]: If you wrap this test in `runBlockingTest` from [Kotlin Coroutines Test](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-blocking-test.html).

It will instantly do the `delay`, and you can accurately do time measurements and comparison.
Not sure if this is really needed, or problematic. Not sure how long this test takes to run.

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 3]
```diff
@@ -1,752 +1,1066 @@
...
-inline fun <M, A> GenTOf<M, A>.fix(): GenT<M, A> =
-    this as GenT<M, A>
-
-typealias Gen<A> = GenT<ForId, A>
-
 /**
- * Datatype that modesl creation of a value based on a seed and a size parameter
+ * A generator wraps a suspend function from ([RandSeed], [Size], [R]) to [Rose].
+ *
+ * Combinators from `arrow-check` are either on present [Gen.Companion] or directly on the generators themselves.
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]:
```suggestion
 * Combinators from `arrow-check` are either present on [Gen.Companion] or directly on the generators themselves.
```

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 4]
```diff
@@ -1,752 +1,1066 @@
...
+ *
+ * This means on every invocation it creates a single result plus all shrunk versions of that result.
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]:
```suggestion
 * This means every invocation creates a single result plus all shrunk versions of that result.
```

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 5]
```diff
@@ -1,752 +1,1066 @@
...
-    override fun BM(): Monad<M> = MM
-    override fun MM(): Monad<GenTPartialOf<M>> = GenT.monad(MM)
-    override fun <A> GenT<M, A>.fromGenT(): Kind<GenTPartialOf<M>, A> = this
-    override fun <A> Kind<GenTPartialOf<M>, A>.toGenT(): GenT<M, A> = fix()
-    override fun <A, B> Kind<GenTPartialOf<M>, A>.map(f: (A) -> B): Kind<GenTPartialOf<M>, B> =
-        fix().genMap(BM(), f)
+/**
+ * Run the generator with an environment [r], effectively removing the need to supply this environment later on.
+ */
+fun <R, A> Gen<R, A>.runEnv(r: R): Gen<Any?, A> = Gen(runGen.compose { (seed, size, _) -> Tuple3(seed, size, r) })
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: Why are we using `Any?` here rather than `Nothing`?

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 6]
```diff
@@ -1,752 +1,1066 @@
...
-                .ensure { it.size >= range.lowerBound(s) }
+/**
+ * Generate a [List] of [A]'s but uses an [IntRange] for convenience.
+ */
+fun <R, A> Gen<R, A>.list(range: IntRange): Gen<R, List<A>> =
+  list(Range.constant(range))
+
+internal fun <R, A> Gen<R, A>.replicate(n: Int): Gen<R, List<A>> =
+  if (n <= 0) Gen.just(emptyList())
+  else (0 until n).toList().fold(Gen.just(emptyList<A>()) as Gen<R, List<A>>) { acc, _ ->
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]:
```suggestion
  else (0 until n).fold<Int, Gen<R, List<A>>>(Gen.just(emptyList<A>())) { acc, _ ->
```

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 7]
```diff
@@ -1,752 +1,1066 @@
...
-        return go(0, emptyMap())
-    }
+/**
+ * Generate a [Map] from a generator that generators key-value pairs.
+ *
+ * The range limits the size of the hashmap.
+ *
+ * This generator will shrink to an empty map and also shrink both keys and values.
+ */
+fun <R, K, A> Gen<R, Pair<K, A>>.hashMap(range: Range<Int>): Gen<R, Map<K, A>> = Gen.sized { s ->
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: Nit: shouldn't we just use `map` here since it reflects what we find in the Std and the type we return here? `Map`. Or should we return `Gen<R, kotlin.collections.HashMap<K, A>>`
 
```suggestion
fun <R, K, A> Gen<R, Pair<K, A>>.map(range: Range<Int>): Gen<R, Map<K, A>> = Gen.sized { s ->
```

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 8]
```diff
@@ -1,752 +1,1066 @@
...
+  }
+
+fun <R, A, B> Gen<R, A>.tupled(other: Gen<R, B>): Gen<R, Pair<A, B>> =
+  map2(other, ::Pair)
+
+fun <R, A, B> Gen.Companion.tupledN(first: Gen<R, A>, second: Gen<R, B>): Gen<R, Pair<A, B>> =
+  first.map2(second, ::Pair)
+
+fun <R, A, B, C> Gen.Companion.tupledN(first: Gen<R, A>, second: Gen<R, B>, third: Gen<R, C>): Gen<R, Triple<A, B, C>> =
+  Gen.mapN(first, second, third) { a, b, c -> Triple(a, b, c) }
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: We're going to deprecate `tupledN` from Arrow since it's just an alias for `mapN` and a lambda reference to a `TupleN` constructor.

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt [hunk 9]
```diff
@@ -1,752 +1,1066 @@
...
+    val (c, defgh) = cdefgh
+    val (d, efgh) = defgh
+    val (e, fgh) = efgh
+    val (f2, gh) = fgh
+    val (g, h) = gh
+    f(a, b, c, d, e, f2, g, h)
+  }
+
+fun <R, A, B> Gen<R, A>.tupled(other: Gen<R, B>): Gen<R, Pair<A, B>> =
+  map2(other, ::Pair)
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: We're going to unify `map2` and `tupled` to `zip` to be uniform with Kotlin's Std.

File: arrow-check/src/main/kotlin/arrow/check/pretty/Diff.kt [hunk 10]
```diff
@@ -44,300 +36,302 @@ import pretty.symbols.space
...
-        is ValueDiffF.ValueD -> ValueDiffF.ValueD(d.l, d.r)
-        is ValueDiffF.TupleD -> ValueDiffF.TupleD(d.vals.map(f))
-        is ValueDiffF.ListD -> ValueDiffF.ListD(d.vals.map(f))
-        is ValueDiffF.Record -> ValueDiffF.Record(d.conName, d.props.map { (k, v) -> k toT f(v) })
-        is ValueDiffF.Cons -> ValueDiffF.Cons(d.consName, d.props.map(f))
-        is ValueDiffF.Same -> ValueDiffF.Same(d.v)
-        is ValueDiffF.ValueDRemoved -> ValueDiffF.ValueDRemoved(d.v)
-        is ValueDiffF.ValueDAdded -> ValueDiffF.ValueDAdded(d.v)
-    }
+internal interface ValueDiffFFunctor : Functor<ForValueDiffF> {
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/pretty/Diff.kt]: The required functionality should probably be inlined into the data type at some point, right?

File: arrow-check/src/main/kotlin/arrow/check/property/Impl.kt [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    }
+  }.let { }
+
+  override suspend fun <R, A> forAllWith(showA: (A) -> Doc<Markup>, env: R, gen: Gen<R, A>): A =
+    suspendCoroutineUninterceptedOrReturn { c ->
+      val labelHere = c.stateStack // save the whole coroutine stack labels
+      val interceptedC = intercepted.get(c)
+      var first = false
+      gen.runEnv(env).flatMap { x: A ->
+        c.stateStack = labelHere
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/property/Impl.kt]: This is a blocker for MPP, why was this needed again? Can we achieve the same semantics with a reasonable API?

File: arrow-check/src/main/kotlin/arrow/check/property/Util.kt [hunk 12]
```diff
@@ -92,87 +173,131 @@ inline class LabelTable(val unLabelTable: String)
...
-                        else acc + mapOf(k2 to v2)
-                    })
-                else acc + mapOf(k to v)
-            }
-        )
+  fun SA(): Semigroup<A>
+  override fun empty(): Coverage<A> = Coverage(emptyMap())
+  override fun Coverage<A>.combine(b: Coverage<A>): Coverage<A> =
+    Coverage(
+      b.unCoverage.toList().fold(unCoverage) { acc, (k, v) ->
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/property/Util.kt]: Nit

```suggestion
      b.unCoverage.entries.fold(unCoverage) { acc, (k, v) ->
```

File: arrow-check/src/main/kotlin/arrow/check/property/Util.kt [hunk 13]
```diff
@@ -92,87 +173,131 @@ inline class LabelTable(val unLabelTable: String)
...
+ * - The constructor
+ * - generated optics
+ * - [PropertyConfig.plus] which accepts any of its properties and overwrites it.
+ *  There are a bunch of smart constructors on the companion object of [PropertyConfig] to help with that.
+ *
+ * @param terminationCriteria Either confidence based termination which checks for coverage
+ * @param maxDiscardRatio Maximum threshold for discarded tests
+ * @param shrinkLimit Maximum shrinking depth. Note this does not refer to shrinking runs, only depth!
+ */
 @optics
```
Comment (nomisRev) [inline arrow-check/src/main/kotlin/arrow/check/property/Util.kt]: Is this being used?

File: arrow-check/src/test/kotlin/arrow/check/gen/GenTest.kt [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+
+    // MonadReader laws
+
+    // Combinators and builders not covered above
+
+    // Individual generators
+    "Gen.long_ with arbitrary range"(
+        PropertyConfig.default() + PropertyConfig.testLimit(10000)
+    ) {
+        val range = forAll(
```
Comment (nomisRev) [inline arrow-check/src/test/kotlin/arrow/check/gen/GenTest.kt]: This is what requires the multi-shot trick, correct?

Another alternative being (didn't work out how I hoped 😅 The DSL is powerful).

```
operator fun <A, B> String.invoke(config: PropertyConfig, genA: Gen<A>, genB: Gen<B>, f: (A, B) -> Unit): Unit = TODO()

fun PropertyTest.test(name: String, l: Long, range: Range, size: Long): Unit {
  val (min, max) = range.bounds(currSize)
  assert(l >= min) { "$name: Generated long was smaller than min".text() }
  assert(l < max) { "$name: Generated long was larger than max".text() }
}

val rangeGen = Gen.mapN(
  Gen.long(Long.MIN_VALUE..Long.MAX_VALUE),
  Gen.long(Long.MIN_VALUE..Long.MAX_VALUE)
) { a, b ->
    if (a >= b) b to a
    else a to b
}.map { (min, max) -> Range.constant(min, max) }

val sizeGen = Gen.sized { Gen.just(it) }

"Gen.long_ with arbitrary range"(
  PropertyConfig.default() + PropertyConfig.testLimit(10000),
  sizeGen,
  // :'(
  rangeGen.flatMap { range -> Gen.long_(range).map { Pair(it, range) } }
  Gen.long_(range)
) { currSize, (l, range) ->
  test("long_", l, range, currSize)
}
```

**Review by nomisRev — APPROVED** [review]
Hey @1Jajen1,
Amazing work! Was hard to properly review since it was so big but managed to get through all of it. Some thoughts and nits but nothing major.

Some concerns about the multi-shot trick, wondering if we can solve it in a different way. This doesn't come with the same downside as `MultiShotCont` or `DelimShotCont` due to the re-wiring of the stack labels, correct? So effects aren't actually re-run since we change the stack labels to the correct position. It's very powerful and neat but ideally, not sure if there are other downsides that we don't know about 😄

File: arrow-check/src/main/kotlin/arrow/check/property/Impl.kt (hunk 11 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/property/Impl.kt]: The only reasonable alternative is to force users to use just one `forAll` per test and combine all generators with `tupledN` or offer an api that does it for them. `forAll` could also function like `flatMap` but that opens up other pitfalls that a) lead to nesting and b) lead to easy to make mistakes.
```kotlin
check {
  val a = forAll(genA)
  val b = forAll(genB)
  val c = forAll(genC)
  ...
}
```
vs
```kotlin
check(tupledN(genA, genB, genC)) { (a, b, c) ->
  ...
}
```

The flat forAll using the suspend hack is by far the nicest syntax, if only it would work on mpp...

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt (hunk 5 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: It avoids a typecast when running it, also `R` has `in` variance so `Any?` is the natural type it converges to isn't it?

File: arrow-check/src/main/kotlin/arrow/check/pretty/Diff.kt (hunk 10 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/pretty/Diff.kt]: Yes, together with all the use of recursion schemes in this file.

File: arrow-check/src/main/kotlin/arrow/check/property/Util.kt (hunk 13 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/property/Util.kt]: The config or the optics accessors? The optics accessors are used a little when updating the config (through api methods) but are not really necessary.

File: arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt (hunk 5 above)
Comment (nomisRev) [reply arrow-check/src/main/kotlin/arrow/check/gen/Gen.kt]: Right, `Any?` for `in` and `Nothing` for `out`.

File: arrow-check/src/main/kotlin/arrow/check/property/Impl.kt (hunk 11 above)
Comment (nomisRev) [reply arrow-check/src/main/kotlin/arrow/check/property/Impl.kt]: > The flat forAll using the suspend hack is by far the nicest syntax, if only it would work on mpp...

100% agreed. The flat syntax is much nicer and more powerful. Although `tupledN` could be hidden from the users API with some boilerplate on the library side writing arity-N overloads... but yes that's not amazing compared to the flat syntax. The flat syntax also comes with major runtime costs due to the stack label resetting.

Sadly, I think that the second option might be better for the time being.

File: arrow-check/src/main/kotlin/arrow/check/property/Util.kt (hunk 13 above)
Comment (nomisRev) [reply arrow-check/src/main/kotlin/arrow/check/property/Util.kt]: The optics. If it's only for little usage I would personally avoid including the dependency.

File: arrow-check/src/test/kotlin/arrow/check/gen/GenTest.kt (hunk 14 above)
Comment (nomisRev) [reply arrow-check/src/test/kotlin/arrow/check/gen/GenTest.kt]: See on-going discussion here: https://github.com/arrow-kt/arrow-check/pull/35#discussion_r559571810

File: arrow-check/src/main/kotlin/arrow/check/property/Impl.kt (hunk 11 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/property/Impl.kt]: > The flat syntax also comes with major runtime costs due to the stack label resetting

This only happens on shrinking (well atm it does for both, but by checking `first` before a reset it could), so normal execution where most tests are expected to succeed doesn't really get hurt from this. Also when a test fails it usually (depending on the test and the shrinker) shrinks in the low 10s-100s of iterations and the printing, diffing and reporting easily dwarves that cost in those cases. Also lets not talk about performance in arrow-check, it's probably fast enough but there are a lot of cases to be improved :see_no_evil: 

Actually since generators are monadic and allow suspend the power of the api is the same if we have one or many `forAll`'s, its just simply not nearly as nice to use...

So lots do a quick pro and cons:
Flat `forAll`:
Pro:
 - **Much nicer syntax**

Cons:
 - **No (full) mpp** (it should be able to support js/jvm through expect/actual)
 - Slows down shrinking
 - More complex implementation which may be a source of errors

Top level `forAll`:
Pro:
 - **Full mpp**
 - Easier to implement
 - Faster when shrinking

Cons:
 - **Annoying to use, monadic use becomes much harder**

If there is a way to (gracefully) fallback to top level gens on native that would be great, since I really don't want to loose this syntax, but also want this to be multi-platform...

File: arrow-check/src/main/kotlin/arrow/check/property/Util.kt (hunk 12 above)
Comment (1Jajen1) [reply arrow-check/src/main/kotlin/arrow/check/property/Util.kt]: Resolved in #41
